crypto/evp/evp_pkey.c

```diff
@@ -79,8 +80,9 @@ EVP_PKEY *EVP_PKCS82PKEY_ex(const PKCS8_PRIV_KEY_INFO *p8, OSSL_LIB_CTX *libctx,
 
     p8_data = encoded_data;
     len = encoded_len;
+    selection = EVP_PKEY_KEYPAIR | EVP_PKEY_KEY_PARAMETERS;
     dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, "DER", "PrivateKeyInfo",
-                                         EVP_PKEY_NONE, 0, libctx, propq);
+                                         NULL, selection, libctx, propq);
```

Comment:
Our documentation of OSSL_DECODER_CTX_new_for_pkey says: `The caller may use the optional input_type, input_struct, keytype and selection to specify what the input is expected to contain.` Optional meaning the selection can be 0 as I understand it. In that case I would expect the decoder to decode _everything_ or _anything_ that is in the data to decode. Why do we need to specify `EVP_PKEY_KEYPAIR | EVP_PKEY_KEY_PARAMETERS` explicitly here and cannot use 0?

Reply:
I'd say that it's reasonable to request a full payload from a PrivateKeyInfo.  In this regard, 0 or `EVP_PKEY_KEYPAIR | EVP_PKEY_KEY_PARAMETERS` really mean the same thing.

Since *this* function is specifically about decoding a PrivateKeyInfo, which is by definition a private key, I see no harm in being specific in this spot.

That being said, the decoders' export function should probably be modified to handle the 0 case, which they don't quite do.
If that should happen in this PR or in another is a good question.  Do you have a preference?

Reply:
It can be a separate PR.